vaxrank/epitope_prediction.py

```diff
@@ -77,24 +86,86 @@ def logistic_score(
 
         return logistic / normalizer
 
-def predict_epitopes(mhc_predictor, protein_fragment):
+# For the given genome/peptide length, compute expected path of cached self-peptides.
+def cache_path(genome, k):
+    cache_dir = user_cache_dir('vaxrank')
+    if not os.path.exists(cache_dir):
+        os.makedirs(cache_dir)
+
+    return os.path.join(cache_dir, '%s_%d_%d.txt' % (
+        genome.species.latin_name, genome.release, k))
+
+def cache_self_peptides_if_needed(genome, ks, force_reload=False):
+    compute_for_ks = []
+    for k in ks:
+        if force_reload or not os.path.exists(cache_path(genome, k)):
+            compute_for_ks.append(k)
+
+    if not compute_for_ks:
+        logger.info("All relevant self-peptides have been cached")
+        return
+
+    kmer_dict = defaultdict(set)
+    logger.info("Computing self-peptides for k=%s", compute_for_ks)
+    for t in genome.transcripts():
+        if not t.is_protein_coding:
+            continue
+        p = t.protein_sequence
+        if not p:
+            continue
+        for k in compute_for_ks:
+            for i in range(len(p) - k + 1):
+                kmer_dict[k].add(p[i:i+k])
+
+    # write out kmer dict data
+    for k, peptides in kmer_dict.items():
+        path = cache_path(genome, k)
+        logger.info("Saving genome %s, k %d to %s", genome, k, path)
+        with open(path, 'w') as f:
+            f.write('\n'.join(peptides))
+
+    logger.info("Done.")
+
+def load_self_peptides(genome, ks, force_reload=False):
+    cache_self_peptides_if_needed(genome, ks, force_reload)
+    logger.info("Loading self-peptides from cache...")
+    kmer_dict = defaultdict(set)
+    for k in ks:
+        with open(cache_path(genome, k)) as f:
+            kmer_dict[k] = set(f.read().split('\n'))
+    logger.info("Done.")
+    return kmer_dict
+
+def predict_epitopes(mhc_predictor, protein_fragment, genome=None):
     results = []
+    self_peptides = load_self_peptides(genome, (8, 9, 10))
```

Comment:
Instead of hard-coded epitope lengths we should use whatever the user picks via a commandline arg.